### PR TITLE
Trace span

### PR DESCRIPTION
**Pull request by selcukcihan**
### Description
Related issue https://linear.app/serverless/issue/SC-356#comment-3690173f
* Enhance/fix trace span parent/child relation
* Add unit tests that validate these relations as well as other functionality as tested in the nodejs sdk.
* Protobuf object and json serialization

### Testing done
Unit tested with
```bash
cd python/packages/sdk
python3 -m venv .venv
source .venv/bin/activate
python3 -m pip install -e .
python3 -m pytest
```

### CODE REVIEW

File: python/packages/sdk/serverless_sdk/sdk/base.py [hunk 1]
```diff
@@ -15,19 +16,28 @@
...
+        return trace.root_span
+
+
 class ServerlessSdk:
     name: Final[str] = __name__
     version: Final[str] = __version__
 
-    trace_spans: Final = ...
+    trace_spans: TraceSpans
     instrumentation: Final = ...
```
Comment (medikoo) [inline python/packages/sdk/serverless_sdk/sdk/base.py:30]: Just wondering, what this `Final = ...` notation means? (looks as syntax error)

File: python/packages/sdk/serverless_sdk/tests/test_trace_span.py [hunk 2]
```diff
@@ -1,152 +1,196 @@
...
+    ), "should expose `trace_id` of a root span"
+    assert type(child_span.id) == str, "should automatically generate `id`"
+    assert child_span.id != root_span.id, "should have a unique `id`"
+    assert (
+        type(child_span.start_time) == int
+    ), "should automatically generate `start_time`"
+    assert (
+        child_span.start_time != root_span.start_time
+    ), "should have a different `start_time`"
+    assert child_span.name == "child", "should expose `name`"
```
Comment (medikoo) [inline python/packages/sdk/serverless_sdk/tests/test_trace_span.py:42]: Do we have a test that confirms that `parent` span is attributed correctly (and that created spans ends in its child spans) ?

**Review by medikoo — COMMENTED** [review]
Thank you @selcukcihan 👍

File: python/packages/sdk/serverless_sdk/sdk/base.py (hunk 1 above)
Comment (selcukcihan) [reply python/packages/sdk/serverless_sdk/sdk/base.py:30]: That's like a placeholder, we'll change it when we actually implement `serverlessSdk.instrumentation`. It's not a syntax error in my environment, maybe you have an old Python version?

File: python/packages/sdk/serverless_sdk/tests/test_trace_span.py (hunk 2 above)
Comment (selcukcihan) [reply python/packages/sdk/serverless_sdk/tests/test_trace_span.py:42]: Can you elaborate a bit more please 🙏  What do you mean by `parent` span being attributed correctly? And also this bit: 
> created spans ends in its child spans

File: python/packages/sdk/serverless_sdk/sdk/base.py (hunk 1 above)
Comment (medikoo) [reply python/packages/sdk/serverless_sdk/sdk/base.py:30]: >  It's not a syntax error in my environment, maybe you have an old Python version?

I just mean that it looks as _syntax error_ and not that it is one.

If I currently will try to access `serverless_sdk.instrumentation` what will be returned? Some empty value? I wonder to what value `...` resolves

File: python/packages/sdk/serverless_sdk/tests/test_trace_span.py (hunk 2 above)
Comment (medikoo) [reply python/packages/sdk/serverless_sdk/tests/test_trace_span.py:42]: When `child_span` is created, we assign a _parent_ to it: https://github.com/serverless/console/blob/052af60e4979974e30386f5808964a210add38d2/python/packages/sdk/serverless_sdk/span/trace.py#L128
So do we confirm in tests that expected _span_ is assigned as `child_span.parent_span` ?

Also do we confirm that `child_span` ends in `parent_span.sub_spans` collection?

Comment (medikoo) [reply python/packages/sdk/serverless_sdk/tests/test_trace_span.py:42]: Those relations are crucial for tracing to work properly, therefore need to be covered with tests

File: python/packages/sdk/serverless_sdk/sdk/base.py (hunk 1 above)
Comment (selcukcihan) [reply python/packages/sdk/serverless_sdk/sdk/base.py:30]: Got it, so it returns an empty object like a stub:
```
>>> from serverless_sdk import serverlessSdk
>>> serverlessSdk.instrumentation
Ellipsis
```

File: python/packages/sdk/serverless_sdk/tests/test_trace_span.py (hunk 2 above)
Comment (selcukcihan) [reply python/packages/sdk/serverless_sdk/tests/test_trace_span.py:42]: Ah I see, yeah good point, I've added these assertions 👍

**Review by medikoo — APPROVED** [review]
Thanks @selcukcihan 👍

**Pull request merged**